core/src/commonMain/kotlin/com/saveourtool/processbuilder/ProcessBuilder.kt

```diff
@@ -0,0 +1,211 @@
+        val cdCmd = if (directory.isNotBlank()) {
+            if (isCurrentOsWindows()) {
+                "cd /d $directory && "
+            } else {
+                "cd $directory && "
+            }
+        } else {
+            ""
+        }
+        // Additionally process command for Windows, it it contain `echo`
+        val commandWithEcho = cdCmd + if (isCurrentOsWindows()) {
+            processCommandWithEcho(command)
+        } else {
+            command
+        }
+        logger.trace { "Modified cmd: $commandWithEcho" }
+        // Finally, make platform dependent adaptations
+        return processBuilderInternal.prepareCmd(commandWithEcho)
+    }
+
+    /**
+     * Log error message and throw exception
+     *
+     * @param errMsg error message
+     * @throws ProcessExecutionException
+     */
+    private fun logErrorAndThrowProcessBuilderException(errMsg: String): Nothing {
+        logger.error { errMsg }
+        throw ProcessExecutionException(errMsg)
+    }
+
+    companion object {
+        /**
+         * Check whether there are exists `echo` commands, and process them, since in Windows
+         * `echo` adds extra whitespaces and newlines. This method will remove them
+         *
+         * @param command command to process
+         * @return unmodified command, if there is no `echo` subcommands, otherwise add parameter `set /p=` to `echo`
+         */
+        @Suppress("ReturnCount")
+        fun processCommandWithEcho(command: String): String {
+            if (!command.contains("echo")) {
+                return command
+            }
+            // Command already contains correct signature.
+            // We also believe not to met complex cases: `echo a; echo | set /p="a && echo b"`
+            val cmdWithoutWhitespaces = command.replace(" ", "")
+            if (cmdWithoutWhitespaces.contains("echo|set")) {
+                return command
+            }
+            if (cmdWithoutWhitespaces.contains("echo\"")) {
+                logger.warn { "You can use echo | set /p\"your command\" to avoid extra whitespaces on Windows" }
+                return command
+            }
+            // If command is complex (have `&&` or `;`), we need to modify only `echo` subcommands
+            val separator = if (command.contains("&&")) {
+                "&&"
+            } else if (command.contains(";")) {
+                ";"
+            } else {
+                ""
+            }
+            val listOfCommands = if (separator != "") command.split(separator) as MutableList<String> else mutableListOf(command)
+            listOfCommands.forEachIndexed { index, cmd ->
+                if (cmd.contains("echo")) {
+                    var newEchoCommand = cmd.trim(' ').replace("echo ", " echo | set /p dummyName=\"")
+                    // Now we need to add closing `"` in proper place
+                    // Despite the fact, that we don't expect user redirections, for out internal tests we use them,
```

Comment:
Also, our logic was build on the idea, that user will not use redirections in his command

But for library, it should be processed somehow

It's necessary to find all places in code with such assumptions and process them (ideally)

But, as it seems to be a very large work, for first step, we should add to README, that we don't process user redirections